cwl/src/main/scala/cwl/CwlInputCoercion.scala

```diff
@@ -0,0 +1,84 @@
+package cwl
+
+import cats.instances.list._
+import cats.syntax.traverse._
+import cats.syntax.validated._
+import cats.syntax.option._
+import common.validation.ErrorOr.ErrorOr
+import common.validation.Validation._
+import cwl.File.FileOrDirectory
+import io.circe.{Json, JsonNumber, JsonObject}
+import shapeless.Coproduct
+import wom.executable.Executable.DelayedCoercionFunction
+import wom.types._
+import wom.values._
+
+// With recursive types we could let circe parse it for us, but until we figure that out just parse it as Json and
+// manually check for File / Directory structures
+private [cwl] object CwlInputCoercion extends Json.Folder[DelayedCoercionFunction] {
+
+  // TODO: This should handle directories
+  def toFile(json: Json): Option[File] = {
+    json.asObject.map( obj => {
+      val values = obj.toMap
+      def getStringValue(key: String) = values.get(key).flatMap(_.asString)
+
+      val secondaryFiles: Option[Array[FileOrDirectory]] = values.get("secondaryFiles").flatMap({ json =>
+        for {
+          array <- json.asArray
+          asFiles = array.flatMap(toFile).toArray.map(Coproduct[FileOrDirectory](_))
+        } yield asFiles
+      })
+
+      File(
+        getStringValue("location"),
+        getStringValue("path"),
+        getStringValue("basename"),
+        getStringValue("dirname"),
+        getStringValue("nameroot"),
+        getStringValue("nameext"),
+        getStringValue("checksum"),
+        values.get("size").flatMap(_.asNumber.flatMap(_.toLong)),
+        secondaryFiles,
+        getStringValue("format"),
+        getStringValue("contents")
+      )
+    })
+  }
+
+  private def simpleCoercion(value: Any)(womType: WomType) = womType.coerceRawValue(value).toErrorOr
+
+  override def onNull = womType => { WomOptionalValue.none(womType).validNel }
+  override def onBoolean(value: Boolean) = simpleCoercion(value)
+  override def onNumber(value: JsonNumber) = {
+    case WomFloatType => WomFloat(value.toDouble).validNel
+    case WomIntegerType => value.toInt.map(WomInteger.apply).toValidNel(s"$value is not a valid Int")
+    case other => other.coerceRawValue(value.toString).toErrorOr
+  }
+  override def onString(value: String) = simpleCoercion(value)
+
+  override def onArray(value: Vector[Json]) = {
+    case womArrayType: WomArrayType =>
+      value.toList
+        .traverse[ErrorOr, WomValue](_.foldWith(CwlInputCoercion).apply(womArrayType.memberType))
```

Comment:
ToL: Here and below I find it slightly easier to reason as `_.foldWith(this)` to realize that this object is being used recursively.